File: src/src/main/resources/mmlspark/_TuneHyperparameters.py
# Copyright (C) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License. See LICENSE in project root for information.


import sys
if sys.version >= '3':
    basestring = str

from pyspark.ml.param.shared import *
from pyspark import keyword_only
from pyspark.ml.util import JavaMLReadable, JavaMLWritable
from pyspark.ml.wrapper import JavaTransformer, JavaEstimator, JavaModel
from pyspark.ml.common import inherit_doc
from mmlspark.Utils import *
from mmlspark.TypeConversionUtils import generateTypeConverter, complexTypeConverter

@inherit_doc
class _TuneHyperparameters(ComplexParamsMixin, JavaMLReadable, JavaMLWritable, JavaEstimator):
    """


    Args:

        evaluationMetric (str): Metric to evaluate models with
        models (object): Estimators to run
        numFolds (int): Number of folds
        numRuns (int): Termination criteria for randomized search
        parallelism (int): The number of models to run in parallel
        paramSpace (object): Parameter space for generating hyperparameters
        seed (long): Random number generator seed (default: 0)
    """

    @keyword_only
    def __init__(self, evaluationMetric=None, models=None, numFolds=None, numRuns=None, parallelism=None, paramSpace=None, seed=0):
        super(_TuneHyperparameters, self).__init__()
        self._java_obj = self._new_java_obj("com.microsoft.ml.spark.TuneHyperparameters")
        self._cache = {}
        self.evaluationMetric = Param(self, "evaluationMetric", "evaluationMetric: Metric to evaluate models with")
        self.models = Param(self, "models", "models: Estimators to run", generateTypeConverter("models", self._cache, complexTypeConverter))
        self.numFolds = Param(self, "numFolds", "numFolds: Number of folds")
        self.numRuns = Param(self, "numRuns", "numRuns: Termination criteria for randomized search")
        self.parallelism = Param(self, "parallelism", "parallelism: The number of models to run in parallel")
        self.paramSpace = Param(self, "paramSpace", "paramSpace: Parameter space for generating hyperparameters")
        self.seed = Param(self, "seed", "seed: Random number generator seed (default: 0)")
        self._setDefault(seed=0)
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        self.setParams(**kwargs)

    @keyword_only
    def setParams(self, evaluationMetric=None, models=None, numFolds=None, numRuns=None, parallelism=None, paramSpace=None, seed=0):
        """
        Set the (keyword only) parameters

        Args:

            evaluationMetric (str): Metric to evaluate models with
            models (object): Estimators to run
            numFolds (int): Number of folds
            numRuns (int): Termination criteria for randomized search
            parallelism (int): The number of models to run in parallel
            paramSpace (object): Parameter space for generating hyperparameters
            seed (long): Random number generator seed (default: 0)
        """
        if hasattr(self, "_input_kwargs"):
            kwargs = self._input_kwargs
        else:
            kwargs = self.__init__._input_kwargs
        return self._set(**kwargs)

    def setEvaluationMetric(self, value):
        """

        Args:

            evaluationMetric (str): Metric to evaluate models with

        """
        self._set(evaluationMetric=value)
        return self


    def getEvaluationMetric(self):
        """

        Returns:

            str: Metric to evaluate models with
        """
        return self.getOrDefault(self.evaluationMetric)


    def setModels(self, value):
        """

        Args:

            models (object): Estimators to run

        """
        self._set(models=value)
        return self


    def getModels(self):
        """

        Returns:

            object: Estimators to run
        """
        return self._cache.get("models", None)


    def setNumFolds(self, value):
        """

        Args:

            numFolds (int): Number of folds

        """
        self._set(numFolds=value)
        return self


    def getNumFolds(self):
        """

        Returns:

            int: Number of folds
        """
        return self.getOrDefault(self.numFolds)


    def setNumRuns(self, value):
        """

        Args:

            numRuns (int): Termination criteria for randomized search

        """
        self._set(numRuns=value)
        return self


    def getNumRuns(self):
        """

        Returns:

            int: Termination criteria for randomized search
        """
        return self.getOrDefault(self.numRuns)


    def setParallelism(self, value):
        """

        Args:

            parallelism (int): The number of models to run in parallel

        """
        self._set(parallelism=value)
        return self


    def getParallelism(self):
        """

        Returns:

            int: The number of models to run in parallel
        """
        return self.getOrDefault(self.parallelism)


    def setParamSpace(self, value):
        """

        Args:

            paramSpace (object): Parameter space for generating hyperparameters

        """
        self._set(paramSpace=value)
        return self


    def getParamSpace(self):
        """

        Returns:

            object: Parameter space for generating hyperparameters
        """
        return self.getOrDefault(self.paramSpace)


    def setSeed(self, value):
        """

        Args:

            seed (long): Random number generator seed (default: 0)

        """
        self._set(seed=value)
        return self


    def getSeed(self):
        """

        Returns:

            long: Random number generator seed (default: 0)
        """
        return self.getOrDefault(self.seed)



    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.TuneHyperparameters"

    @staticmethod
    def _from_java(java_stage):
        module_name=_TuneHyperparameters.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".TuneHyperparameters"
        return from_java(java_stage, module_name)

    def _create_model(self, java_model):
        return _TuneHyperparametersModel(java_model)


class _TuneHyperparametersModel(ComplexParamsMixin, JavaModel, JavaMLWritable, JavaMLReadable):
    """
    Model fitted by :class:`_TuneHyperparameters`.

    This class is left empty on purpose.
    All necessary methods are exposed through inheritance.
    """

    @classmethod
    def read(cls):
        """ Returns an MLReader instance for this class. """
        return JavaMMLReader(cls)

    @staticmethod
    def getJavaPackage():
        """ Returns package name String. """
        return "com.microsoft.ml.spark.TuneHyperparametersModel"

    @staticmethod
    def _from_java(java_stage):
        module_name=_TuneHyperparametersModel.__module__
        module_name=module_name.rsplit(".", 1)[0] + ".TuneHyperparametersModel"
        return from_java(java_stage, module_name)

File: dotnetSanityCheck/dotnetClasses/TuneHyperparameters.cs
// Copyright (C) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See LICENSE in project root for information.


using System;
using System.Collections.Generic;
using System.Linq;
using System.Reflection;
using Microsoft.Spark.ML.Feature.Param;
using Microsoft.Spark.Interop;
using Microsoft.Spark.Interop.Ipc;
using Microsoft.Spark.ML.Feature;
using Microsoft.Spark.Sql;
using Microsoft.Spark.Sql.Types;
using Microsoft.Spark.Interop.Internal.Java.Util;
using Microsoft.Spark.Utils;

using SynapseML.Dotnet.Utils;
using Synapse.ML.Automl;

namespace Synapse.ML.Automl
{
    /// <summary>
    /// <see cref="TuneHyperparameters"/> implements TuneHyperparameters
    /// </summary>
    public class TuneHyperparameters : JavaEstimator<TuneHyperparametersModel>, IJavaMLWritable, IJavaMLReadable<TuneHyperparameters>
    {
        private static readonly string s_className = "com.microsoft.azure.synapse.ml.automl.TuneHyperparameters";

        /// <summary>
        /// Creates a <see cref="TuneHyperparameters"/> without any parameters.
        /// </summary>
        public TuneHyperparameters() : base(s_className)
        {
        }

        /// <summary>
        /// Creates a <see cref="TuneHyperparameters"/> with a UID that is used to give the
        /// <see cref="TuneHyperparameters"/> a unique ID.
        /// </summary>
        /// <param name="uid">An immutable unique ID for the object and its derivatives.</param>
        public TuneHyperparameters(string uid) : base(s_className, uid)
        {
        }

        internal TuneHyperparameters(JvmObjectReference jvmObject) : base(jvmObject)
        {
        }

        /// <summary>
        /// Sets evaluationMetric value for <see cref="evaluationMetric"/>
        /// </summary>
        /// <param name="evaluationMetric">
        /// Metric to evaluate models with
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetEvaluationMetric(string value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setEvaluationMetric", (object)value));
        
        /// <summary>
        /// Sets models value for <see cref="models"/>
        /// </summary>
        /// <param name="models">
        /// Estimators to run
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetModels(IEstimator<object>[] value)
        {
            var arrayList = new ArrayList(SparkEnvironment.JvmBridge);
            foreach (var v in value)
            {
                arrayList.Add(v);
            }
            return WrapAsTuneHyperparameters(Reference.Invoke("setModels", (object)arrayList));
        }
        
        /// <summary>
        /// Sets numFolds value for <see cref="numFolds"/>
        /// </summary>
        /// <param name="numFolds">
        /// Number of folds
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetNumFolds(int value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setNumFolds", (object)value));
        
        /// <summary>
        /// Sets numRuns value for <see cref="numRuns"/>
        /// </summary>
        /// <param name="numRuns">
        /// Termination criteria for randomized search
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetNumRuns(int value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setNumRuns", (object)value));
        
        /// <summary>
        /// Sets parallelism value for <see cref="parallelism"/>
        /// </summary>
        /// <param name="parallelism">
        /// The number of models to run in parallel
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetParallelism(int value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setParallelism", (object)value));
        
        /// <summary>
        /// Sets paramSpace value for <see cref="paramSpace"/>
        /// </summary>
        /// <param name="paramSpace">
        /// Parameter space for generating hyperparameters
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetParamSpace(object value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setParamSpace", value));
        
        /// <summary>
        /// Sets seed value for <see cref="seed"/>
        /// </summary>
        /// <param name="seed">
        /// Random number generator seed
        /// </param>
        /// <returns> New TuneHyperparameters object </returns>
        public TuneHyperparameters SetSeed(long value) =>
            WrapAsTuneHyperparameters(Reference.Invoke("setSeed", (object)value));

        
        /// <summary>
        /// Gets evaluationMetric value for <see cref="evaluationMetric"/>
        /// </summary>
        /// <returns>
        /// evaluationMetric: Metric to evaluate models with
        /// </returns>
        public string GetEvaluationMetric() =>
            (string)Reference.Invoke("getEvaluationMetric");
        
        
        /// <summary>
        /// Gets models value for <see cref="models"/>
        /// </summary>
        /// <returns>
        /// models: Estimators to run
        /// </returns>
        public IEstimator<object>[] GetModels()
        {
            JvmObjectReference[] jvmObjects = (JvmObjectReference[])Reference.Invoke("getModels");
            IEstimator<object>[] result = new IEstimator<object>[jvmObjects.Length];
            Dictionary<string, Type> classMapping = JvmObjectUtils.ConstructJavaClassMapping(
                typeof(JavaPipelineStage),
                "s_className");
            for (int i=0; i < jvmObjects.Length; i++)
            {
                if (JvmObjectUtils.TryConstructInstanceFromJvmObject(
                    jvmObjects[i],
                    classMapping,
                    out IEstimator<object> instance))
                {
                    result[i] = instance;
                }
            }
            return result;
        }
        
        
        /// <summary>
        /// Gets numFolds value for <see cref="numFolds"/>
        /// </summary>
        /// <returns>
        /// numFolds: Number of folds
        /// </returns>
        public int GetNumFolds() =>
            (int)Reference.Invoke("getNumFolds");
        
        
        /// <summary>
        /// Gets numRuns value for <see cref="numRuns"/>
        /// </summary>
        /// <returns>
        /// numRuns: Termination criteria for randomized search
        /// </returns>
        public int GetNumRuns() =>
            (int)Reference.Invoke("getNumRuns");
        
        
        /// <summary>
        /// Gets parallelism value for <see cref="parallelism"/>
        /// </summary>
        /// <returns>
        /// parallelism: The number of models to run in parallel
        /// </returns>
        public int GetParallelism() =>
            (int)Reference.Invoke("getParallelism");
        
        
        /// <summary>
        /// Gets paramSpace value for <see cref="paramSpace"/>
        /// </summary>
        /// <returns>
        /// paramSpace: Parameter space for generating hyperparameters
        /// </returns>
        public object GetParamSpace() => Reference.Invoke("getParamSpace");
        
        
        /// <summary>
        /// Gets seed value for <see cref="seed"/>
        /// </summary>
        /// <returns>
        /// seed: Random number generator seed
        /// </returns>
        public long GetSeed() =>
            (long)Reference.Invoke("getSeed");

        /// <summary>Fits a model to the input data.</summary>
        /// <param name="dataset">The <see cref="DataFrame"/> to fit the model to.</param>
        /// <returns><see cref="TuneHyperparametersModel"/></returns>
        override public TuneHyperparametersModel Fit(DataFrame dataset) =>
            new TuneHyperparametersModel(
                (JvmObjectReference)Reference.Invoke("fit", dataset));

        /// <summary>
        /// Loads the <see cref="TuneHyperparameters"/> that was previously saved using Save(string).
        /// </summary>
        /// <param name="path">The path the previous <see cref="TuneHyperparameters"/> was saved to</param>
        /// <returns>New <see cref="TuneHyperparameters"/> object, loaded from path.</returns>
        public static TuneHyperparameters Load(string path) => WrapAsTuneHyperparameters(
            SparkEnvironment.JvmBridge.CallStaticJavaMethod(s_className, "load", path));
        
        /// <summary>
        /// Saves the object so that it can be loaded later using Load. Note that these objects
        /// can be shared with Scala by Loading or Saving in Scala.
        /// </summary>
        /// <param name="path">The path to save the object to</param>
        public void Save(string path) => Reference.Invoke("save", path);
        
        /// <returns>a <see cref="JavaMLWriter"/> instance for this ML instance.</returns>
        public JavaMLWriter Write() =>
            new JavaMLWriter((JvmObjectReference)Reference.Invoke("write"));
        
        /// <returns>an <see cref="JavaMLReader"/> instance for this ML instance.</returns>
        public JavaMLReader<TuneHyperparameters> Read() =>
            new JavaMLReader<TuneHyperparameters>((JvmObjectReference)Reference.Invoke("read"));

        private static TuneHyperparameters WrapAsTuneHyperparameters(object obj) =>
            new TuneHyperparameters((JvmObjectReference)obj);

        
    }
}

        